File: services/glossary-service/internal/llmbudget/llmbudget.go
// Package llmbudget is glossary-service's call-profile registry — one row per kind of LLM
// call it makes, mirroring composition-service's `app/llm_budget.py` and translation's
// `app/llm_budget.py`. The SDK owns the mechanism; the per-operation facts live with the
// service that knows them.
//
// # What was wrong
//
// Both of glossary's LLM call sites (`runDocExtractor`, `runPlanner`) built a
// `llm.StreamRequest` with NO MaxTokens and never looked at `res.FinishReason` — a field the
// Go SDK populates from the terminal `done` event and both sites discarded.
//
// No cap is a legitimate choice (see `OutputKind.MIRROR`: for a translation the model's
// natural stop IS the bound). It is NOT legitimate here, and the difference is the whole
// point of the kind vocabulary: both of these calls parse JSON. A clipped JSON object does
// not come back short, it comes back UNPARSEABLE — `truncation_is_fatal` in
// contracts/llm-budget.contract.json.
//
// # Why the repair round made it worse
//
// On a parse failure each site runs ONE repair round that re-prompts with
//
//	"Your previous output was invalid: <parse error>"
//
// If the real cause was truncation, that sentence is a lie the model cannot recover from: it
// is told it produced malformed syntax when it actually produced correct syntax that got cut
// off, so it "fixes" grammar it never got wrong — against the same absent budget. The
// likeliest way for a model to satisfy both the complaint and the limit is to emit a SHORTER
// list. That parses cleanly, reports success, and silently drops entities.
//
// So a truncation must be named as a truncation, never fed to the repair round.
package llmbudget

import "fmt"

// TruncatedFinishReason is the wire value provider-registry reports when the model stopped
// because it hit the cap. Pinned against contracts/llm-budget.contract.json by the test in
// this package.
const TruncatedFinishReason = "length"

// Profile is one row: what the call may produce, and what a clip costs.
type Profile struct {
	// MaxTokens is the output cap. 0 would mean "omit — let the model decide", which is the
	// platform-wide sentinel; no row here uses it, because every glossary LLM call parses
	// JSON and an uncapped structured call is precisely the failure documented above.
	MaxTokens int
	// TruncationIsFatal ⇒ a `finish_reason == "length"` DESTROYS the result rather than
	// shortening it, so the caller must treat it as an error and must not "repair" it.
	TruncationIsFatal bool
	Why               string
}

// TokensPerItem is the worst-case JSON cost of ONE emitted row — a kind, a name, up to a
// dozen attribute key/value pairs, a scope label and punctuation. Taken from the Python
// SDK's `_TOKENS_PER_ITEM` so the three registries size the same shape the same way.
const TokensPerItem = 220

// Profiles is the registry, keyed by the function the call lives in.
//
// A cap does NOT prevent a clip; a cap BELOW the legitimate output causes one. Both of these
// calls were previously uncapped, so any fixed number here can only introduce a failure that
// did not exist — which is why the row for a caller-sized output is the omit sentinel and the
// row for a bounded one is derived from that bound.
var Profiles = map[string]Profile{
	// The caller passes `maxCandidates` (up to maxDocExtractCandidates = 200). At
	// TokensPerItem that is ~44k, so ANY flat cap that fits a small model truncates a large
	// document — the first version of this row used 4096, roughly 18 candidates against a
	// declared maximum of 200, on a call that had no cap at all before. The output length
	// here is dictated by the request, exactly like a translation's is dictated by its
	// source, so the model's natural stop is the right bound and `Truncated()` is the guard.
	"doc_extract": {
		MaxTokens:         omitSentinel,
		TruncationIsFatal: true,
		Why: "size is caller-driven (maxCandidates ≤ 200 ≈ 44k tokens) — a flat cap can only " +
			"clip a large doc; the finish_reason check is the guard",
	},
	"doc_extract_repair": {
		MaxTokens:         omitSentinel,
		TruncationIsFatal: true,
		Why:               "the repair re-emits the SAME object, so it inherits the same reasoning",
	},
	// A plan is BOUNDED — loreweave_mcp.MaxPlanOps = 50, enforced at the planner ("over the
	// cap is an error at the planner, never a silent truncation"). So the budget is knowable:
	// 50 × TokensPerItem = 11k. 12000 is the number plan-forge already runs against real
	// models in this repo, and it is ≥ the bound, so no valid plan can hit it.
	"action_plan": {
		MaxTokens:         12000,
		TruncationIsFatal: true,
		Why:               "a validated plan is bounded by MaxPlanOps=50 (~11k tokens); 12000 clears it",
	},
	"action_plan_repair": {
		MaxTokens:         12000,
		TruncationIsFatal: true,
		Why:               "the repair re-emits the SAME object, so it needs the same room",
	},
}

// omitSentinel — 0 means "no cap; drop the field from the wire", the platform-wide
// convention (contracts/llm-budget.contract.json; the Go SDK's `,omitempty` implements it).
const omitSentinel = 0

// For returns the row for code. Unknown codes panic rather than defaulting: a silent
// fallback would re-create the unattributed budget this registry exists to remove, and the
// codes are compile-time constants at their call sites, so a bad one is a programming error
// caught by the package's own test — never a runtime surprise on a user's request.
func For(code string) Profile {
	p, ok := Profiles[code]
	if !ok {
		panic(fmt.Sprintf("unknown glossary call profile %q — add a row to llmbudget.Profiles "+
			"rather than passing a literal at the call site", code))
	}
	return p
}

// MaxTokensFor is the value to put on the wire for code.
func MaxTokensFor(code string) int { return For(code).MaxTokens }

// Truncated reports whether the model stopped because it ran out of budget.
func Truncated(finishReason string) bool { return finishReason == TruncatedFinishReason }

// TruncationError describes a clip in the terms a caller (and the agent reading the tool
// error) can act on. Deliberately NOT phrased as a malformed-output error: that
// misattribution is the bug this package exists to prevent.
//
// The limit is reported as the PROVIDER's when this row sends no cap of its own — otherwise
// the message would name 0 and read as nonsense. An uncapped call that still clipped was
// bounded by the provider (Anthropic substitutes 8192 for a missing cap) or by the context
// window, and the actionable advice differs.
func TruncationError(code string, outputTokens int) error {
	limit := MaxTokensFor(code)
	where := fmt.Sprintf("its output limit of %d tokens", limit)
	advice := fmt.Sprintf("raise the %q budget in internal/llmbudget", code)
	if limit == omitSentinel {
		where = "the PROVIDER's own output limit (this call sends no cap of its own — " +
			"Anthropic substitutes 8192 for a missing one)"
		advice = "use a model with a larger output window"
	}
	return fmt.Errorf(
		"the model's response was cut off at %s (finish_reason=%q) after %d tokens, so the "+
			"JSON is incomplete — this is a capacity limit, not a malformed answer. Retry "+
			"with a smaller input (fewer/shorter passages), or %s",
		where, TruncatedFinishReason, outputTokens, advice)
}

File: services/lore-enrichment-service/app/llm_budget.py
"""lore-enrichment-service's call-profile registry — one row per kind of LLM call it makes.

Same split as composition-service's registry and `PASS_REGISTRY` before it: the SDK
(`loreweave_llm.budget.call_budget`) owns the MECHANISM — what an output kind means, how a
target is sized, how reasoning eats the same allowance — and the per-operation facts live here,
next to the code that knows them.

Why this file exists at all
---------------------------
This service had two hand-rolled `POST /internal/llm/stream` bodies, and the budget gate did
not parse that shape. It said so in its PASS line rather than pretending otherwise, which was
honest — and still hid a live defect for as long as the surface stayed unscanned: the eval
judge in `app/eval/judge_binding.py` sent **no `max_tokens` key at all**. Uncapped, on a call
whose consumer parses strict JSON.

An excluded surface with an accurate label is still an excluded surface.

Two rows, and they are not the same shape
-----------------------------------------
`generate_gap_completion` is an ADOPTION: it had a real number (4000) and this row must never
resolve below it, which the registry test checks against `was`.

`eval_judge_usefulness` is a NARROWING: it had no cap, so there is no previous number for a
no-downgrade test to compare against and `was=None` says so. A row that recorded `was=0` would
satisfy `budget_for(code) >= was` for every conceivable value — a test that cannot fail,
dressed as coverage. The exemption instead carries `narrowing_why`, and the registry test
requires it to be non-empty, so the escape hatch has to state its own reason.
"""
from __future__ import annotations

from dataclasses import dataclass

from loreweave_llm.budget import CallBudget, OutputKind, call_budget

__all__ = ["CallProfile", "PROFILES", "budget_for", "max_tokens_for", "profile_for"]


@dataclass(frozen=True)
class CallProfile:
    kind: OutputKind
    #: The measured minimum for this call.
    floor: int
    #: The budget ACTUALLY IN USE before this row existed. `None` ⇒ the call was UNCAPPED, so
    #: this row narrows rather than adopts and the no-downgrade assertion does not apply.
    #: A `None` here obliges `narrowing_why`.
    was: int | None
    why: str = ""
    #: Required exactly when `was is None`: why capping a previously uncapped call is safe.
    narrowing_why: str = ""


#: code → profile. A code is the FUNCTION the call lives in.
PROFILES: dict[str, CallProfile] = {
    # ── the gap-completion prose seam (app/generation/complete.py).
    #
    # PROSE, so `language` is a signal the kind actually reads — the caller holds it on
    # `StrategyContext.profile.language`, and a CJK completion needs materially more tokens
    # for the same prose than a Latin-script one.
    "generate_gap_completion": CallProfile(
        OutputKind.PROSE, floor=4000, was=4000,
        why="one entity's enriched profile/bio prose; matches wiki's per-article budget"),

    # ── the C15 eval judge ensemble (app/eval/judge_binding.py).
    #
    # STRUCTURED and not VERDICT, which is the judgement in this row. The kind is decided by
    # what TRUNCATION COSTS, not by how the call is described, and `parse_judge_verdict`
    # accepts a JSON object or nothing: a clipped `{"verdict": "hi` is not a shorter answer,
    # it is no answer, and the proposal silently drops out of that judge's denominator. That
    # is STRUCTURED's definition — and it is why this row's adoption also puts a
    # finish_reason check at the call site, which the contract requires of a fatal kind.
    "eval_judge_usefulness": CallProfile(
        OutputKind.STRUCTURED, floor=4096, was=None,
        why="one `{verdict, ...}` object per proposal",
        narrowing_why=(
            "The call was UNCAPPED. The judge's whole output is a small JSON verdict object "
            "and STRUCTURED's 4096 floor is roughly two orders of magnitude more than one "
            "needs, so the cap cannot clip a well-formed answer; what it does bound is a "
            "judge that ignores the rubric and rambles — which today burns tokens and then "
            "fails to parse anyway. Uncapped was never the intent, only the default."),
    ),
}


def profile_for(code: str) -> CallProfile:
    try:
        return PROFILES[code]
    except KeyError:
        raise KeyError(
            f"no call profile for {code!r} — add a row to app/llm_budget.py rather than "
            f"passing a literal max_tokens"
        ) from None


def budget_for(code: str, *, target: int | None = None, language: str | None = None,
               reasoning=None, context_length: int | None = None) -> CallBudget:
    """Resolve `code`'s budget, threading whatever per-call signal the caller holds."""
    p = profile_for(code)
    return call_budget(
        p.kind, target=target, language=language, reasoning=reasoning,
        context_length=context_length, floor=p.floor,
    )


def max_tokens_for(code: str, **kw) -> int:
    """The `max_tokens` value for the wire."""
    return budget_for(code, **kw).max_output_tokens
